### DISCUSSION TITLE
Figure out how to properly use fstrim

### DISCUSSION THREAD

**Discussion by agrover** [stratis-storage/project#555] [Ideas]
Talk to Brian or Dennis on VDO team on how to use fstrim in an optimal fashion.

**Comment by leseb** [comment]
Is this still a topic? I didn't find `fstrim` being used anywhere in the codebase.
From my experience, it's discouraged to mount the fs with a discard option. Running scheduled `fstrim` jobs is preferred. The only use case where it is appropriate is if we have a fast device with fast discards and FTL/firmware that is known to behave reasonably well. Also, this might not be true as the device ages so periodic trims are encouraged.

Hope this helps.

**Reply by jbaublitz** [reply]
We had previously discussed two options. One would be to run fstrim periodically for the user in the background. The other is to provide a D-Bus API for running fstrim at the filesystem level. Do you have thoughts on either of those?

**Reply by leseb** [reply]
Both solutions are appealing, also I think it depends on which layer we want to operate, typically trimming has two areas: SSD/LUNs and thin-provisioned storage. Stratis could have a pool that has both capabilities, it runs over SSDs and is thin-provisioned. I think there is another scenario with an SSD pool and a thick provisioned fs but I don't think Stratis supports thick provisioning.

### Pool of SSDs or LUNs case

Trimming is particularly important during removal operations, for instance, if an fs is deleted then it's important to issue discard on the dm to reclaim the blocks on the SSD. LVM has a [flag](https://github.com/lvmteam/lvm2/blob/master/conf/example.conf.in#L371-L383) to control this.

### Thin provisioned fs
Running a periodic fstrim in the background should be configurable by the user IMO and on a per fs basis. The D-Bus API is important to let Stratis handle the trim instead of the user having to know which command to execute. Both options are not mutually exclusive and an admin could decide to issue a trim on an fs manually. There is more to consider too, the fs level is good because it's granular, but a wrapper for a pool that would issue the trim on all the fs might be interesting.
Also, Stratis would need to check whether the fs is already mounted with discard or not so that we don't need to do anything.

### Other things to consider

* Filesystem creation sending discard, see https://github.com/stratis-storage/stratisd/issues/3178
* Also when open an encrypted block with cryptsetup, there is a "--allow-discard" option
